api/controllers/grades.go

```diff
@@ -6,283 +6,202 @@ import (
 
-		professor := (first_name != "" || last_name != "")
+	groupGradeDistributionStage := bson.D{
+		{Key: "$group", Value: bson.D{
+			{Key: "_id", Value: "$_id.academic_session"},
+			{Key: "grade_distribution", Value: bson.D{{Key: "$push", Value: "$grades"}}},
+		}},
+	}
 
-		lookupSectionsStage := bson.D{
-			{Key: "$lookup", Value: bson.D{
-				{Key: "from", Value: "sections"},
-				{Key: "localField", Value: "sections"},
-				{Key: "foreignField", Value: "_id"},
-				{Key: "as", Value: "sections"},
-			}},
+	switch {
+	case prefix != "" && number == "" && sectionNumber == "" && !professor:
+		// Filter on Course
+		collection = courseCollection
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"subject_prefix": prefix}}})
+
+	case prefix != "" && number != "" && sectionNumber == "" && !professor:
+		// Filter on Course
+		collection = courseCollection
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"subject_prefix": prefix, "course_number": number}}})
+
+	case prefix != "" && number != "" && sectionNumber != "" && !professor:
+		// Filter on Course then Section
+		collection = courseCollection
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"subject_prefix": prefix, "course_number": number}}})
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"sections.section_number": sectionNumber}}})
+
+	case prefix == "" && number == "" && sectionNumber == "" && professor:
+		// Filter on Professor
+		collection = professorCollection
+		match := bson.M{}
+		if firstName != "" {
+			match["first_name"] = firstName
+		}
+		if lastName != "" {
+			match["last_name"] = lastName
 		}
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: match}})
 
-		unwindSectionsStage := bson.D{{Key: "$unwind", Value: bson.D{{Key: "path", Value: "$sections"}}}}
+	case prefix != "" && professor:
+		// Filter on Section by Matching Course and Professor IDs
+		collection = sectionCollection
 
-		projectGradeDistributionStage := bson.D{
-			{Key: "$project", Value: bson.D{
-				{Key: "_id", Value: "$sections.academic_session.name"},
-				{Key: "grade_distribution", Value: "$sections.grade_distribution"},
-			}},
+		var profIDs []primitive.ObjectID
+		match := bson.M{"professors": bson.M{}}
+		if firstName != "" {
+			match["first_name"] = firstName
 		}
-
-		projectGradeDistributionWithSectionsStage := bson.D{
-			{Key: "$project", Value: bson.D{
-				{Key: "_id", Value: "$academic_session.name"},
-				{Key: "grade_distribution", Value: "$grade_distribution"},
-			}},
+		if lastName != "" {
+			match["last_name"] = lastName
 		}
-
-		unwindGradeDistributionStage := bson.D{
-			{Key: "$unwind", Value: bson.D{
-				{Key: "path", Value: "$grade_distribution"},
-				{Key: "includeArrayIndex", Value: "ix"},
-			}},
+		cursor, err = professorCollection.Find(ctx, bson.D{{Key: "$match", Value: match}})
```

Comment:
None of this will work properly, since you'll end up with a query that looks like:
```mql
{
 $match: {
   "professors": {},
   "first_name": first_name,
   "last_name": last_name 
 }
}
```
which isn't a valid query.